master/internal/db/migrations.go

```diff
@@ -111,8 +116,117 @@ func ensureMigrationUpgrade(tx *pg.Tx) error {
 	return nil
 }
 
+func (db *PgDB) readDBCodeAndCheckIfDifferent(dbCodeDir string) (map[string]string, bool, error) {
+	files, err := os.ReadDir(dbCodeDir)
+	if err != nil {
+		return nil, false, fmt.Errorf("reading '%s' directory for database views: %w", dbCodeDir, err)
+	}
+
+	allCode := ""
+	fileNamesToSQL := make(map[string]string)
+	for _, f := range files {
+		if filepath.Ext(f.Name()) != ".sql" {
+			continue
+		}
+
+		filePath := filepath.Join(dbCodeDir, f.Name())
+		b, err := os.ReadFile(filePath) //nolint: gosec // We trust dbCodeDir.
+		if err != nil {
+			return nil, false, fmt.Errorf("reading view definition file '%s': %w", filePath, err)
+		}
+		fileNamesToSQL[f.Name()] = string(b)
+		allCode += string(b)
+	}
+
+	// I didn't want to get into deciding when to apply database or code or not but integration
+	// tests make it really hard to not do this.
+	hash := sha256.Sum256([]byte(allCode))
+	ourHash := hex.EncodeToString(hash[:])
+
+	// Check if the views_and_triggers_hash table exists. If it doesn't return that we need to create db code.
+	var tableExists bool
+	if err = db.sql.QueryRow(
+		"SELECT EXISTS(SELECT 1 FROM information_schema.tables WHERE table_name = 'views_and_triggers_hash')").
+		Scan(&tableExists); err != nil {
+		return nil, false, fmt.Errorf("checking views_and_triggers_hash exists: %w", err)
+	}
+	if !tableExists {
+		return fileNamesToSQL, true, nil
+	}
+
+	// Check if our hashes match. If they do we can just return we don't need to do anything.
+	var databaseHash string
+	if err := db.sql.QueryRow("SELECT hash FROM views_and_triggers_hash").Scan(&databaseHash); err != nil {
+		return nil, false, fmt.Errorf("getting hash from views_and_triggers_hash: %w", err)
+	}
+	if databaseHash == ourHash {
+		return fileNamesToSQL, false, nil
+	}
+
+	// Update our hash and return we need to create views and triggers.
+	if _, err := db.sql.Exec("UPDATE views_and_triggers_hash SET hash = $1", ourHash); err != nil {
+		return nil, false, fmt.Errorf("updating our database hash: %w", err)
+	}
+	return fileNamesToSQL, false, nil
+}
+
+func (db *PgDB) addDBCode(fileNamesToSQL map[string]string) error {
+	if err := db.withTransaction("determined database views", func(tx *sqlx.Tx) error {
+		for filePath, sql := range fileNamesToSQL {
+			if _, err := tx.Exec(sql); err != nil {
+				return fmt.Errorf("running database view file '%s': %w", filePath, err)
+			}
+		}
+
+		return nil
+	}); err != nil {
+		return fmt.Errorf("adding determined database views: %w", err)
+	}
+
+	return nil
+}
+
+func (db *PgDB) dropDBCode() error {
+	// SET search_path since the ALTER DATABASE ... SET SEARCH_PATH won't apply to this connection
```

Comment:
im confused by this comment - can you elaborate?

Reply:
oh I think this comment is out of date, see https://github.com/determined-ai/determined/pull/9302/files/3f6fe3ef6d5bed31c711ad171757e22ac090f6ee#diff-5b35744fbe2a9f5695da280e46848172deb3cdaedada8395015b997c1cb87578R209